module/os/linux/zfs/zfs_vfsops.c

```diff
@@ -2275,6 +2274,7 @@ zfs_resume_fs(zfsvfs_t *zfsvfs, dsl_dataset_t *ds)
 	    zp = list_next(&zfsvfs->z_all_znodes, zp)) {
 		err2 = zfs_rezget(zp);
 		if (err2) {
+			zpl_d_drop_aliases(ZTOI(zp));
```

Comment:
After giving this some more thought I think we need to always drop all of the aliases, even when the `zfs_rezget()` succeeds.  Otherwise, we'll incorrectly handle the case when the file does still exist after the rollback, and a hard link to that files gets removed by the rollback, but that dentry doesn't get correctly dropped.  

The `z_is_stale` field can be removed, and the comment above should be updated.  How about:

```
        /* 
         * Attempt to re-establish all the active inodes with their dbufs.
         * If zfs_rezget() fails, then we unhash the inode to prevent a
         * collision if a new inode is allocated with the same inode/object
         * number.  All dentries for the inode are dropped from the cache
         * since they may no longer be valid and a fresh lookup is required.
         */
```

Reply:
No problem, would you have any hints & recommendations wrt/ testing environment?

I'd try some ACL changes between the snapshot and rollback too; is it OK to add and remove users in the test suite?

So I'll try:
- changing file metadata and see if that works
- hardlinks
- what else, please? :)

Reply:
> I'd try some ACL changes between the snapshot and rollback too; is it OK to add and remove users in the test suite?

Great.  Yes, you can add and remove users in the ZTS with the `add_user` and `del_user` helper functions.  Take a look at `tests/functional/userquota/` tests for an example of this.

I think we should verify all of the following for correctness after an online rollback.

* File types: normal files, directories, hardlinks, symlinks, `O_TMPFILE`.
* Metadata: permissions, group, owner, xattrs (sa and diretory)
* Rollback invalidates negative dentries
* Rollback with open file handles
* Rollback with open file handles of unlinked files.

Additionally, I think we would get a lot of value stressing the code by including a test case which runs for a minute or two and randomly performs snapshots and rollbacks of an active filesystem.  In particular, this would be a good way to test open file handles and unlinks are handled properly.  Related to exactly this I opened #9739 with a proposed fix.

Reply:
I've been trying to figure out, how to do the stress test part, but I can't come up with an elegant solution, which I could really call 'stress' test.

For one, writing a multithreaded beast in userspace C is something I'd rather avoid, but even if I bite the bullet, what should the threads do, so that it be considered stress testing stuff and not just repeating the same workload over?

So I'm thinking it's best to admit, that I'm stuck and ask for help, so that we can move on with this :)

Maybe let's skip the 'stress' test part, only do various permutations of ^ what you listed, perhaps repeat that process a few times and check the status of things in-between the steps - opening a file and leaving it open should be possible directly in the Korn shell. That's a fallback solution, so that we get at least somewhere, to move the whole OverlayFS thing forward in reasonable time frame. Would that work?

Reply:
That sounds like a reasonable path forward to me.  We can always add additional test case as we come up with specific scenarios.

Reply:
@didrocks says I might be able to get some help from the Docker guys with writing the tests - for which I'm sorry, but I can't seem to find the time ... bugs are popping up in weird places and we desperately need to finish the kernel @ our infra. Sh*t.

I can understand the pain of not having the overlay2 Docker storage driver available...